lib/MaRC/VirtualImage.h

```diff
@@ -27,120 +27,354 @@
 
 #include "MaRC/SourceImage.h"
 
+#include <limits>
+#include <cmath>
+
 
 namespace MaRC
 {
-
-  /**
-   * @class VirtualImage
-   *
-   * @brief Base class for virtual source images.
-   *
-   * Data from virtual images are computed at run-time rather than
-   * retrieved from static sources such as images stored on a
-   * filesystem.
-   */
-  class VirtualImage : public SourceImage
-  {
-  public:
-
-    /// Constructor.
-    /**
-     * @param s Linear scaling coefficient applied to computed data.
-     * @param o Linear offset value applied to all (scaled) computed
-     *          data.
-     */
-    VirtualImage(double s = 1, double o = 0);
-
-    /// Retrieve data from virtual image.
     /**
-     * Retrieve data from virtual image and apply configured data
-     * transformations, if any.  Raw data is computed/retrieved from
-     * the read_data_i() template method.
+     * @brief Integer data scale and offset calculation functor.
      *
-     * @param lat  Bodycentric (e.g. planetocentric) latitude in radians.
-     * @param lon  Longitude in radians.
+     * Determine the best scale and offset parameters applied to
+     * floating point data read from source images to retain as many
+     * significant digits as possible when storing that data in
+     * integer typed maps.
      *
-     * @param[out] data Data retrieved from image.
+     * @see MaRC::scale_and_offset()
      *
-     * @return @c true - Data retrieved, @c false - No data retrieved.
-     *
-     * @see read_data_i()
+     * @internal This functor is not part of public MaRC library API.
      */
-    virtual bool read_data(double lat, double lon, double & data) const;
+    template <typename T>
+    struct scale_and_offset_impl
+    {
+        bool operator()(double min,
+                        double max,
+                        double & scale,
+                        double & offset) const
+        {
+            constexpr double T_lowest = std::numeric_limits<T>::lowest();
+            constexpr double T_max    = std::numeric_limits<T>::max();
+            constexpr double type_range = T_max - T_lowest;
+
+            double const data_range = max - min;
+
+            if (data_range < 0 || data_range > type_range) {
+                // min > max or can't fit all data into desired type T.
+                return false;
+            }
+
+            int const exponent =
+                static_cast<int>(std::numeric_limits<T>::digits10)
+                - static_cast<int>(std::log10(data_range));
 
-  private:
+            scale = std::pow(10, exponent);
+
```

Comment:
This offset calculation is not correct for minimum and maximum physical data values that are not symmetrical, such as longitudes in the range [0, 360].  Fixed for bug #42.